compiler/rmi/kotlinr/src/org/jetbrains/kotlin/rmi/kotlinr/KotlinCompilerClient.kt

```diff
@@ -18,246 +18,376 @@ package org.jetbrains.kotlin.rmi.kotlinr
+            if (locked) {
+                lock?.release()
+                channel?.close()
+                lockFile.delete()
+                locked = false
             }
+        }
 
-            val daemon = connectToCompileService(compilerId, daemonLaunchingOptions, daemonOptions, System.out, autostart = !clientOptions.stop, checkId = !clientOptions.stop)
+        private val channel = if (locked) RandomAccessFile(lockFile, "rw").channel else null
+        private val lock = channel?.lock()
 
-            if (daemon == null) {
-                if (clientOptions.stop) println("No daemon found to shut down")
-                else throw Exception("Unable to connect to daemon")
+        synchronized private fun acquireLockFile(lockFile: File): Boolean {
+            if (lockFile.createNewFile()) return true
+            try {
+                // attempt to delete if file is orphaned
+                if (lockFile.delete() && lockFile.createNewFile())
+                    return true // if orphaned file deleted assuming that the probability of
             }
-            else when {
-                clientOptions.stop -> {
-                    println("Shutdown the daemon")
-                    daemon.shutdown()
-                    println("Daemon shut down successfully")
-                }
-                else -> {
-                    println("Executing daemon compilation with args: " + filteredArgs.joinToString(" "))
-                    val outStrm = RemoteOutputStreamServer(System.out)
-                    try {
-                        val memBefore = daemon.getUsedMemory() / 1024
-                        val startTime = System.nanoTime()
-                        val res = daemon.remoteCompile(filteredArgs.toArrayList().toTypedArray(), outStrm, CompileService.OutputFormat.PLAIN)
-                        val endTime = System.nanoTime()
-                        println("Compilation result code: $res")
-                        val memAfter = daemon.getUsedMemory() / 1024
-                        println("Compilation time: " + TimeUnit.NANOSECONDS.toMillis(endTime - startTime) + " ms")
-                        println("Used memory $memAfter (${"%+d".format(memAfter - memBefore)} kb)")
-                    }
-                    finally {
-                        outStrm.disconnect()
-                    }
-                }
+            catch (e: IOException) {
+                // Ignoring it - assuming it is another client process owning it
             }
+            var attempts = 0L
+            while (lockFile.exists() && attempts++ * COMPILE_DAEMON_STARTUP_LOCK_TIMEOUT_CHECK_MS < COMPILE_DAEMON_STARTUP_LOCK_TIMEOUT_MS) {
+                Thread.sleep(COMPILE_DAEMON_STARTUP_LOCK_TIMEOUT_CHECK_MS)
+            }
+            if (lockFile.exists())
+                throw IOException("Timeout waiting the release of the lock file '${lockFile.absolutePath}")
+
+            return lockFile.createNewFile()
```

Comment:
Can't we replace it with something like that:
```kotlin
do {
    if (lockFile.createNewFile()) return true
    Thread.sleep(...)
} while (!timeout)

throw timeout exception
```

It is better because the function has only two cases: return or exception. Now we have true/false/exception